### DISCUSSION TITLE
Any way to decrease time to target power?

### DISCUSSION THREAD

**Discussion by holycacao** [doudar/SmartSpin2k#554] [Q&A]
Loving my SS2K. One thing I can't quite figure out is how to decrease the time it takes to reach a higher target power (while maintaining the same cadence). Here's a screenshot of a recent ride in TrainerRoad with power peaks and valleys.

![image](https://github.com/doudar/SmartSpin2k/assets/6627625/61f29175-1a9a-4bd1-a906-0936d2887133)

I notice the SS2K will substantially spin down in one continuous turn of the knob to reach a lower power target, but for higher power, it uses multiple smaller steps to ratchet up the resistance, often taking 10-15 seconds. In this ride (Renoso), the peaks are 40 seconds long, so taking 10-15 seconds isn't terrible, but the peak interval can be as short as 20 seconds in other rides, which results in almost the entire interval being spent ratcheting up to the target power. You can see below that I've set the ERG Sensitivity to 10 (this helps) and Stepper Motor Speed to the max of 3500 (this helps some too).

![ss2k-settings-01](https://github.com/doudar/SmartSpin2k/assets/6627625/27e7af44-a49b-4d4e-aec3-284dc7241f6a)
![ss2k-settings-02](https://github.com/doudar/SmartSpin2k/assets/6627625/2d22e960-0e6d-44d4-a348-a9b8486a78b7)
![ss2k-settings-03](https://github.com/doudar/SmartSpin2k/assets/6627625/f87f4541-a74d-4989-bd2e-4c259da6017f)

Any ideas?

**Comment by eMadman** [comment]
You've made the right move by increasing the erg multiplier and motor speed.  

For now, you can help it along by spinning up your cadence a few seconds before the interval, similar to how they recommend in Zwift.

There is something being worked on that might help with this.  It's a couple of weeks out.

**Reply by holycacao** [reply]
Thanks, I'll give that a shot. Just out of curiosity, do you know why SS2K is seemingly more conservative when increasing resistance to reach a higher target power vs when it decreases it to a lower one?

**Reply by eMadman** [reply]
The device creates a lookup table for resistance.  The table is calculated at 50 watt increments using 90rpm as the baseline cadence.  It is possible that it doesn't have a lookup point captured near your interval target.   You can try to force this by doing a free ride.  Hold a 90rpm cadence and slowly increase resistance with the shifter till you're somewhere above the upper interval target.  Hold the resistance and cadence for 3-5 seconds at each increment.  This should create some lookup values at the upper end.  

The firmware update being worked on right now will create the lookup table for additional cadence ranges and hopefully help further:
![image](https://github.com/doudar/SmartSpin2k/assets/24726844/a2d5ff0d-48cb-4f23-84af-bb3ab0ce8269)

**Reply by holycacao** [reply]
Is it possible for me to view the generated resistance chart like you've posted above?

**Reply by eMadman** [reply]
This particular chart is part of the work in progress.  This side of it is likely to end up in the config app.  


You can view the current iteration of the resistance table data by enabling logs and capturing them.  Instructions here: https://github.com/doudar/SmartSpin2k/wiki/Viewing-logs-via-UDP

Upload a text file output from a short ride.  This will also help me see what's been captured for your resistance table.

Just a heads up that the setting to enable UDP logging is on theweb portal and not in the current version of the mobile app.  You'll need to connect the smartspin2k to your home wifi.  Use the app to connect it to wifi - it works better for most people.  Then navigate to http://smartspin2k.local to get to the settings screen.

**Reply by holycacao** [reply]
I did a ride first as a baseline (`ss2k-ride.log`), but admittedly not the best ride for testing this because it didn't have extreme peaks and valleys (it was the next one on my training calendar).

Then tried to generate a look-up table (`ss2k-lu-table.log`). It took me a while to figure out that I needed to quit the "train with your favorite app" mode on my C7 and instead switch to manual ride to get the shifters to work. So, the actual data is probably near the end of the file.

I'll be paying attention during the next couple of rides to see if anything changes.

[ss2k-ride.log](https://github.com/user-attachments/files/15971388/ss2k-ride.log)
[ss2k-lu-table.log](https://github.com/user-attachments/files/15971390/ss2k-lu-table.log)

**Comment by eMadman** [comment]
From your lookup table ride, here's the table that was generated:

```
[125922][E](TTable): |     1|     5|    11|    15|    21|    27|    32|     0|     0|     0|     0|     0|     0|     0|     0|     0|     0|     0|     0|     0|
[125925][E](TTable): |-36845| -3071|  -718|  2141|  3460|  5133|  6000|     0|     0|     0|     0|     0|     0|     0|     0|     0|     0|     0|     0|     0|
```

Along the top row of the table, you see torque.  This is calculated as Torque=(watts*9.5488)/cadence

The use of torque allows more flexibility in the rider's cadence.  The second row is the stepper motor position that hit that target output. 

The 27nm value you generated was while you were putting out around 260 watts.  The 32 value was around 300 watts.


Now looking at your Erg ride:
```
[2544525][E](TTable): |     1|     0|    11|    16|    20|    25|     0|     0|     0|     0|     0|     0|     0|     0|     0|     0|     0|     0|     0|     0|
[2544528][E](TTable): |-38475|     0| -3065|   836|  2485|  3712|     0|     0|     0|     0|     0|     0|     0|     0|     0|     0|     0|     0|     0|     0|
```
Highest value is 25nm... Approx 240 watts at 90rpm.  Still within the range of your targets for your intervals.  My best guess is cadence could be a factor from your original report?  @doudar is more fit to give an answer on this - he'll be back later this weekend.  

I'm quite busy this week but I'll see if I can find some time to test the work in progress firmware to see if it throws any glaring errors.  You might be a good candidate to help test it in the wild.

**Reply by holycacao** [reply]
Thanks for the info! My next ride has high, short (20 sec) peaks, so I'll report back here how it went and include logs as well.

**Reply by holycacao** [reply]
New ride with high, short peaks. I noticed that if I actually _slowed_ my cadence some before the peak, the resistance increased faster and in bigger segments (or at least it felt that), getting to the desired output faster. Increasing cadence before the peak didn't seem to do the trick.
![image](https://github.com/doudar/SmartSpin2k/assets/6627625/f1197bf7-ef40-42f8-b851-f4876f1c82e2)

[ss2k-ride-2.log](https://github.com/user-attachments/files/16013741/ss2k-ride-2.log)
